### PR TITLE
Fix #16654 and partially fix #16583

### LINKED ISSUE #16583: Regression on Tuple.Union when using Tuple.toList
## Compiler version

Worked on `3.2.1`
Fails on `3.2.2-RC2`

## Minimized code

```Scala
  val ll0: Tuple3["one", "two", "three"] = constValueTuple[("one", "two", "three")]
  val ll1 = constValueTuple[("one", "two", "three")].toList
  val ll3: List["one" | ("two" | ("three" | Nothing))] = constValueTuple[("one", "two", "three")].toList
  val ll4: List["one" | ("two" | "three")] = constValueTuple[("one", "two", "three")].toList

  inline def labels[Labels <: Tuple](using ev: Tuple.Union[Labels] <:< String): List[String] = 
    val tmp = constValueTuple[Labels].toList
    ev.substituteCo( tmp )

  val strings = labels[("one", "two", "three")]
  // returns List(one, two, three)
  println(strings.mkString("<", ", ", ">"))
```

## Output

```scala
error] -- [E007] Type Mismatch Error: /home/hmf/VSCodeProjects/sploty/meta/src/data/Meta1.scala:46:53 
[error] 46 |  val ll1 = constValueTuple[("one", "two", "three")].toList
[error]    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]    |Found:    List[
[error]    |  Tuple.Union[(?1 : (("one" : String), ("two" : String), ("three" : String)))]
[error]    |]
[error]    |Required: List[("one" : String) | (("two" : String) | (("three" : String) | Nothing))]
[error]    |
[error]    |where:    ?1 is an unknown value of type (("one" : String), ("two" : String), ("three" : String))
[error]    |----------------------------------------------------------------------------
[error]    | Explanation (enabled by `-explain`)
[error]    |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
[error]    |
[error]    | Tree: scala.compiletime.constValueTuple[Tuple3["one".type, "two".type, "three".type]].
[error]    |   toList
[error]    | I tried to show that
[error]    |   List[
[error]    |   Tuple.Union[(?1 : (("one" : String), ("two" : String), ("three" : String)))]
[error]    | ]
[error]    | conforms to
[error]    |   List[("one" : String) | (("two" : String) | (("three" : String) | Nothing))]
[error]    | but the comparison trace ended with `false`:
[error]    |
[error]    |   ==> List[   Tuple.Union[(?1 : (("one" : String), ("two" : String), ("three" : String)))] ]  <:  List[("one" : String) | (("two" : String) | (("three" : String) | Nothing))]
[error]    |     ==> Tuple.Union[(?1 : (("one" : String), ("two" : String), ("three" : String)))]  <:  ("one" : String) | (("two" : String) | (("three" : String) | Nothing))
[error]    |       ==> Tuple.Union[(?1 : (("one" : String), ("two" : String), ("three" : String)))]  <:  Nothing
[error]    |         ==> lub(Nothing, ("one" : String), canConstrain=false, isSoft=true)
[error]    |         <== lub(Nothing, ("one" : String), canConstrain=false, isSoft=true) = ("one" : String)
[error]    |         ==> lub(Nothing, (("two" : String), ("three" : String)), canConstrain=false, isSoft=true)
[error]    |         <== lub(Nothing, (("two" : String), ("three" : String)), canConstrain=false, isSoft=true) = (("two" : String), ("three" : String))
[error]    |         ==> lub(Nothing, ("two" : String), canConstrain=false, isSoft=true)
[error]    |         <== lub(Nothing, ("two" : String), canConstrain=false, isSoft=true) = ("two" : String)
[error]    |         ==> lub(Nothing, ("three" : String) *: EmptyTuple.type, canConstrain=false, isSoft=true)
[error]    |         <== lub(Nothing, ("three" : String) *: EmptyTuple.type, canConstrain=false, isSoft=true) = ("three" : String) *: EmptyTuple.type
[error]    |         ==> lub(Nothing, ("three" : String), canConstrain=false, isSoft=true)
[error]    |         <== lub(Nothing, ("three" : String), canConstrain=false, isSoft=true) = ("three" : String)
[error]    |         ==> lub(Nothing, EmptyTuple.type, canConstrain=false, isSoft=true)
[error]    |         <== lub(Nothing, EmptyTuple.type, canConstrain=false, isSoft=true) = EmptyTuple.type
[error]    |         ==> Nothing  <:  ("three" : String) in frozen constraint
[error]    |         <== Nothing  <:  ("three" : String) in frozen constraint = true
[error]    |         ==> lub(<notype>, <notype>, canConstrain=false, isSoft=true)
[error]    |         <== lub(<notype>, <notype>, canConstrain=false, isSoft=true) = <notype>
[error]    |         ==> lub(<notype>, <notype>, canConstrain=false, isSoft=true)
[error]    |         <== lub(<notype>, <notype>, canConstrain=false, isSoft=true) = <notype>
[error]    |         ==> lub(Nothing, ("one" : String), canConstrain=false, isSoft=true)
[error]    |         <== lub(Nothing, ("one" : String), canConstrain=false, isSoft=true) = ("one" : String)
[error]    |         ==> lub(Nothing, (("two" : String), ("three" : String)), canConstrain=false, isSoft=true)
[error]    |         <== lub(Nothing, (("two" : String), ("three" : String)), canConstrain=false, isSoft=true) = (("two" : String), ("three" : String))
[error]    |         ==> lub(Nothing, ("two" : String), canConstrain=false, isSoft=true)
[error]    |         <== lub(Nothing, ("two" : String), canConstrain=false, isSoft=true) = ("two" : String)
[error]    |         ==> lub(Nothing, ("three" : String) *: EmptyTuple.type, canConstrain=false, isSoft=true)
[error]    |         <== lub(Nothing, ("three" : String) *: EmptyTuple.type, canConstrain=false, isSoft=true) = ("three" : String) *: EmptyTuple.type
[error]    |         ==> lub(Nothing, ("three" : String), canConstrain=false, isSoft=true)
[error]    |         <== lub(Nothing, ("three" : String), canConstrain=false, isSoft=true) = ("three" : String)
[error]    |         ==> lub(Nothing, EmptyTuple.type, canConstrain=false, isSoft=true)
[error]    |         <== lub(Nothing, EmptyTuple.type, canConstrain=false, isSoft=true) = EmptyTuple.type
[error]    |         ==> ("one" : String) | (("two" : String) | (("three" : String) | Nothing))  <:  Nothing
[error]    |           ==> String  <:  Nothing in frozen constraint
[error]    |           <== String  <:  Nothing in frozen constraint = false
[error]    |           ==> ("one" : String)  <:  Nothing
[error]    |             ==> String  <:  Nothing (left is approximated)
[error]    |             <== String  <:  Nothing (left is approximated) = false
[error]    |           <== ("one" : String)  <:  Nothing = false
[error]    |         <== ("one" : String) | (("two" : String) | (("three" : String) | Nothing))  <:  Nothing = false
[error]    |       <== Tuple.Union[(?1 : (("one" : String), ("two" : String), ("three" : String)))]  <:  Nothing = false
[error]    |     <== Tuple.Union[(?1 : (("one" : String), ("two" : String), ("three" : String)))]  <:  ("one" : String) | (("two" : String) | (("three" : String) | Nothing)) = false
[error]    |   <== List[   Tuple.Union[(?1 : (("one" : String), ("two" : String), ("three" : String)))] ]  <:  List[("one" : String) | (("two" : String) | (("three" : String) | Nothing))] = false

```

## Expectation

That it will compile on the latest Scala compiler version.

### LINKED ISSUE #16654: Regression in gekomad/itto-csv
https://scala3.westeurope.cloudapp.azure.com/job/buildCommunityProject/38542/

Scala version: 3.2.2-RC2

Bisect points to 46e3d773ce1cae594196dc984356e2629d3fd74a

Reproducer: 
```scala
def toCsvFlat[A <: Product](a: A)(using m: scala.deriving.Mirror.ProductOf[A]) = {
  def flatTuple(any: Any): Tuple = any match
    case p: Product => p.productIterator.map(flatTuple).foldLeft(EmptyTuple: Tuple)(_ ++ _)
    case a          => Tuple1(a)
    
  val tuple = flatTuple(Tuple.fromProductTyped(a)).toList
}
```

Output: 

```scala
-- [E007] Type Mismatch Error: /workspace/bisect/test.scala:6:51 ---------------
6 |  val tuple = flatTuple(Tuple.fromProductTyped(a)).toList
  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |Found:    List[Tuple.Union[(?1 : Tuple)]]
  |Required: List[
  |  Tuple match {
  |    case EmptyTuple => Nothing
  |    case h *: t => h | scala.Tuple.Fold[t, Nothing, [x, y] =>> x | y]
  |  }
  |]
  |
  |where:    ?1 is an unknown value of type Tuple
  |
  |
  |Note: a match type could not be fully reduced:
  |
  |  trying to reduce  Tuple.Union[(?1 : Tuple)]
  |  trying to reduce  scala.Tuple.Fold[(?1 : Tuple), Nothing, [x, y] =>> x | y]
  |  failed since selector  (?1 : Tuple)
  |  does not match  case EmptyTuple => Nothing
  |  and cannot be shown to be disjoint from it either.
  |  Therefore, reduction cannot advance to the remaining case
  |
  |    case h *: t => h | scala.Tuple.Fold[t, Nothing, [x, y] =>> x | y]
  |
```

### PR DESCRIPTION
**Pull request by mbovel**
Fixes #16654 and part of #16583.

### CODE REVIEW

File: library/src/scala/Tuple.scala
Comment (sjrd) [inline library/src/scala/Tuple.scala:16]: This is not backward tasty compatible, unfortunately.

**Review by sjrd — CHANGES_REQUESTED** [review]

File: library/src/scala/Tuple.scala
Comment (mbovel) [reply library/src/scala/Tuple.scala:16]: Does it mean it should be merged after the next minor release only, or that we can't do that at all?

Comment (sjrd) [reply library/src/scala/Tuple.scala:16]: We cannot do it at all.

Comment (mbovel) [reply library/src/scala/Tuple.scala:16]: Too bad 🥲
We have the same problem for `Tuple.map` by the way: https://github.com/lampepfl/dotty/issues/15992#issuecomment-1243700382.

Comment (bishabosha) [reply library/src/scala/Tuple.scala:16]: how can we sensibly improve API's known to be broken? to we need to add an `.api2` method to Tuple which returns an object with all the fixed APIs?

Comment (sjrd) [reply library/src/scala/Tuple.scala:16]: The only true way is not to introduce broken APIs in the first place ... which means more testing when they are introduced. Of course that's too late for what's already published.

It might be possible to pull off something once `@binaryAPI` is introduced. As we could keep the broken version as `@binaryAPI private[Tuple]` and introduce the fixed version as public.

Comment (bishabosha) [reply library/src/scala/Tuple.scala:16]: > As we could keep the broken version as @binaryAPI private[Tuple] and introduce the fixed version as public.

I thought when inlining that would fail the accessibility check, as it would already be resolved to the specific overload?

Comment (sjrd) [reply library/src/scala/Tuple.scala:16]: Accessibility from tasty is already relaxed to a degree.

Comment (mbovel) [reply library/src/scala/Tuple.scala:16]: Suggestion from Martin: deprecate `Tuple.toList` and replace it with a new `Tuple.toIterator` that has a type parameter.

Not sure what to do for `Tuple.map`.

**Comment by mbovel** [comment]
Following the discussion above, I rolled back the change to the signature of `Tuple.toList`.

As a result, while all other tests and minimization of #16583 are fixed by the `compareAtoms` and `ApproximatingTypeMap` fixes, `labels` can still not compile:

```scala
inline def labels[Labels <: Tuple](using ev: Tuple.Union[Labels] <:< String): List[String] =
  val tmp = constValueTuple[Labels].toList
  ev.substituteCo( tmp )
```

```
-- [E007] Type Mismatch Error: tests/pos/16583.scala:11:19 ---------------------
11 |  ev.substituteCo( tmp )
   |                   ^^^
   | Found:    (tmp : List[Any])
   | Required: List[Tuple.Union[Labels]]
   |
   | where:    Labels is a type in method labels with bounds <: Tuple
   |
   |
   | Note: a match type could not be fully reduced:
   |
   |   trying to reduce  Tuple.Union[Labels]
   |   trying to reduce  scala.Tuple.Fold[Labels, Nothing, [x, y] =>> x | y]
   |   failed since selector Labels
   |   does not match  case EmptyTuple => Nothing
   |   and cannot be shown to be disjoint from it either.
   |   Therefore, reduction cannot advance to the remaining case
   |
   |     case h *: t => h | scala.Tuple.Fold[t, Nothing, [x, y] =>> x | y]
   |
   | longer explanation available when compiling with `-explain`
```

This makes sense. Not sure how we could fix this without changing the signature of `Tuple.toList`.

Writing the expected type of `tmp` explicitly:

```scala
val tmp: List[Tuple.Union[Labels]] = (constValueTuple[Labels]: Labels).toList
```

```
-- [E007] Type Mismatch Error: tests/pos/16583.scala:10:73 ---------------------
10 |  val tmp: List[Tuple.Union[Labels]] = (constValueTuple[Labels]: Labels).toList
   |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |Found:    List[Tuple.Union[(?1 : Labels)]]
   |Required: List[Tuple.Union[Labels]]
   |
   |where:    ?1     is an unknown value of type Labels
   |          Labels is a type in method labels with bounds <: Tuple
   |
   |
   |Note: a match type could not be fully reduced:
   |
   |  trying to reduce  Tuple.Union[(?1 : Labels)]
   |  trying to reduce  scala.Tuple.Fold[(?1 : Labels), Nothing, [x, y] =>> x | y]
   |  failed since selector (?1 : Labels)
   |  does not match  case EmptyTuple => Nothing
   |  and cannot be shown to be disjoint from it either.
   |  Therefore, reduction cannot advance to the remaining case
   |
   |    case h *: t => h | scala.Tuple.Fold[t, Nothing, [x, y] =>> x | y]
   |
   | longer explanation available when compiling with `-explain`
```

**Comment by mbovel** [comment]
Note: I kept individual commits for now to ease reviews, but will rebase to 2 commits (`compareAtoms` and `ApproximatingTypeMap` fixes) before merging.

**Comment by sjrd** [comment]
I have dismissed my earlier review since the API does not change anymore. However, I am woefully unqualified to sign off on the changes that do remain in the PR, so I'll refrain from posting a new review.

**Comment by mbovel** [comment]
Failing community test:

```
[error] -- [E057] Type Mismatch Error: /__w/dotty/dotty/community-build/community-projects/perspective/dotty/examples/src/main/scala/perspective/examples/testing.scala:144:29 
[8876](https://github.com/lampepfl/dotty/actions/runs/5143935279/jobs/9259577653#step:8:8877)
[error] 144 |val res = Cat("Garfield").foo("name")
[8877](https://github.com/lampepfl/dotty/actions/runs/5143935279/jobs/9259577653#step:8:8878)
[error]     |                             ^
[8878](https://github.com/lampepfl/dotty/actions/runs/5143935279/jobs/9259577653#step:8:8879)
[error]     |Type argument ("name" : String) | Nothing does not conform to lower bound Any
[8879](https://github.com/lampepfl/dotty/actions/runs/5143935279/jobs/9259577653#step:8:8880)
[error]     |---------------------------------------------------------------------------
[8880](https://github.com/lampepfl/dotty/actions/runs/5143935279/jobs/9259577653#step:8:8881)
[error]     |Inline stack trace
[8881](https://github.com/lampepfl/dotty/actions/runs/5143935279/jobs/9259577653#step:8:8882)
[error]     |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
[8882](https://github.com/lampepfl/dotty/actions/runs/5143935279/jobs/9259577653#step:8:8883)
[error]     |This location contains code that was inlined from hkdGeneric.scala:87
[8883](https://github.com/lampepfl/dotty/actions/runs/5143935279/jobs/9259577653#step:8:8884)
[error]  87 |      constValueTuple[m.MirroredElemLabels].toList.toSet.asInstanceOf[Set[String]]
[8884](https://github.com/lampepfl/dotty/actions/runs/5143935279/jobs/9259577653#step:8:8885)
[error]     |                                                   ^
[8885](https://github.com/lampepfl/dotty/actions/runs/5143935279/jobs/9259577653#step:8:8886)
[error]      ---------------------------------------------------------------------------
[8886](https://github.com/lampepfl/dotty/actions/runs/5143935279/jobs/9259577653#step:8:8887)
[error]     |
[8887](https://github.com/lampepfl/dotty/actions/runs/5143935279/jobs/9259577653#step:8:8888)
[error]     | longer explanation available when compiling with `-explain`
```

Code: https://github.com/dotty-staging/perspective/blob/365383df6a7b0c2eeb81742f479aa56269a4b557/dotty/examples/src/main/scala/perspective/examples/testing.scala#L137-L144.

At first sight, that's another case wher a type application argument is a skolem of an abstract type, resulting in the application being approximated. This is similar to the `labels` example above.

**Comment by smarter** [comment]
There's too much going on in this PR I think, can we split the AllowLambdaWildcardApply change in its own PR so we can evaluate it separately?

**Comment by dwijnand** [comment]
Found a way to make it all work, changing less:  #18043

**Comment by mbovel** [comment]
Let's try to get https://github.com/lampepfl/dotty/pull/18043 merged first, and then to address the `AllowLambdaWildcardApply` change in a separate PR as suggested.

**Pull request closed without merging**
